### PR TITLE
add a CLI command to get fission url and launch UI

### PR DESCRIPTION
**Pull request by yqf3139**
- issue #190  instead of use the url directly, the url is set to env variables
  manually
- issue #203

### CODE REVIEW

**Comment by soamvasani** [comment]
Thanks @yqf3139 ! 

[@life1347 was also going to work on #190, so you two might want to talk... in the future it will help to comment on an issue when you start working on it, to avoid duplication of efforts.]

If we use the [apiserver proxy](https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-services/#manually-constructing-apiserver-proxy-urls), we don't need to expose the controller service at all, and we can take advantage of apiserver authentication too.  This will be more reliable and secure than our current set up.

File: fission/common.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ func fatal(msg string) {
 func getClient(serverUrl string) *client.Client {
 
 	if len(serverUrl) == 0 {
-		fatal("Need --server or FISSION_URL set to your fission server.")
+		fatal("Need --server or FISSION_URL set to your fission server. Or try source <(fission get-env-variables)")
```
Comment (soamvasani) [inline fission/common.go]: We should eliminate this fatal error, and get the url from the API. Why make the user set a var that we know how to get automatically?

File: fission/main.go [hunk 2]
```diff
@@ -17,11 +17,105 @@ limitations under the License.
...
 
 import (
+	"errors"
+	"fmt"
 	"os"
+	"path/filepath"
 
+	"github.com/skratchdot/open-golang/open"
 	"github.com/urfave/cli"
+	"k8s.io/client-go/1.5/kubernetes"
```
Comment (soamvasani) [inline fission/main.go]: Would you mind splitting the k8s-client stuff into a separate file in this package?

File: fission/main.go [hunk 3]
```diff
@@ -17,11 +17,105 @@ limitations under the License.
...
+		// TODO: find the port via name
+		ip, err3 := getNodeIP(k8sClientset)
+		if err3 != nil {
+			panic(err3)
+		}
+		return fmt.Sprintf("http://%v:%v",
+			ip, controller.Spec.Ports[0].NodePort)
+	case v1.ServiceTypeLoadBalancer:
+		return fmt.Sprintf("http://%v:%v",
+			controller.Status.LoadBalancer.Ingress[0].IP, controller.Spec.Ports[0].Port)
```
Comment (soamvasani) [inline fission/main.go]: Sometimes the IP isn't set, and there's a hostname instead (seems to happen on AWS).

File: fission/main.go [hunk 4]
```diff
@@ -124,6 +218,35 @@ func main() {
 				return nil
 			},
 		},
+		{
+			Name:  "get-env-variables",
```
Comment (soamvasani) [inline fission/main.go]: Suggest removing this.

File: fission/common.go (hunk 1 above)
Comment (yqf3139) [inline fission/common.go]: We can get the server var only when the k8s config is present and valid. So keeping the environment variable is a good idea?

File: fission/main.go (hunk 4 above)
Comment (yqf3139) [inline fission/main.go]: Yes. Current implementation is not easy to use. But we should find a way to cache the fission url or we need to fetch the url every time cli runs.

File: fission/common.go (hunk 1 above)
Comment (soamvasani) [inline fission/common.go]: Yes we should allow the user to set the env var to override the default. But we shouldn't fail if it's unset.

File: fission/main.go (hunk 4 above)
Comment (soamvasani) [inline fission/main.go]: How slow is it to fetch the url every time? And btw, if we switch to using the apiserver proxy, we won't have to talk to the api to get the url.

Comment (yqf3139) [inline fission/main.go]: Less than two seconds, depends on where your k8s cluster is.
For apiserver proxy, we still need to know the k8s master ip from the k8s local config. I haven't dig into that.

Comment (life1347) [inline fission/main.go]: Two seconds is a little bit long to me. How about write related information into local config file (/home/.fission/config)?

Comment (yqf3139) [inline fission/main.go]: @soamvasani 
Local config file is good to to me. What is your opinion?

**Comment by soamvasani** [comment]
closing in favor of #431

**Pull request closed without merging**